src/DataGrid.tsx

```diff
@@ -154,52 +175,83 @@ function DataGrid<R, K extends keyof R, SR>({
+  onRowExpandToggle,
+  onSelectedCellChange,
+  onSelectedCellRangeChange,
   onSelectedRowsChange,
   ...props
 }: DataGridProps<R, K, SR>, ref: React.Ref<DataGridHandle>) {
-  const [columnWidths, setColumnWidths] = useState<ReadonlyMap<string, number>>(() => new Map());
-  const [scrollLeft, setScrollLeft] = useState(0);
-  const [gridWidth, setGridWidth] = useState(0);
+  /**
+   * refs
+   * */
   const gridRef = useRef<HTMLDivElement>(null);
-  const headerRef = useRef<HTMLDivElement>(null);
-  const viewportWidth = (width || gridWidth) - 2; // 2 for border width;
+  const lastSelectedRowIdx = useRef(-1);
 
-  const columnMetrics = useMemo(() => {
-    if (viewportWidth <= 0) return null;
+  /**
+   * states
+   */
+  const [eventBus] = useState(() => new EventBus());
+  const [gridWidth, setGridWidth] = useState(0);
+  const [scrollTop, setScrollTop] = useState(0);
+  const [scrollLeft, setScrollLeft] = useState(0);
+  const [columnWidths, setColumnWidths] = useState<ReadonlyMap<string, number>>(() => new Map());
+
+  /**
+   * computed values
+   */
+  const viewportWidth = (width || gridWidth) - 2; // 2 for border width;
 
+  const { columns, lastFrozenColumnIndex, totalColumnWidth } = useMemo(() => {
     return getColumnMetrics<R, SR>({
-      columns,
+      columns: rawColumns,
       minColumnWidth,
       viewportWidth,
       columnWidths,
       defaultFormatter
     });
-  }, [columnWidths, columns, defaultFormatter, minColumnWidth, viewportWidth]);
+  }, [columnWidths, rawColumns, defaultFormatter, minColumnWidth, viewportWidth]);
 
   const [colOverscanStartIdx, colOverscanEndIdx] = useMemo((): [number, number] => {
-    if (!columnMetrics) {
-      return [0, 0];
-    }
-
-    return getHorizontalRangeToRender({
-      columnMetrics,
+    return getHorizontalRangeToRender(
+      columns,
+      lastFrozenColumnIndex,
+      viewportWidth,
       scrollLeft
-    });
-  }, [columnMetrics, scrollLeft]);
+    );
+  }, [scrollLeft, columns, lastFrozenColumnIndex, viewportWidth]);
 
   const viewportColumns = useMemo((): readonly CalculatedColumn<R, SR>[] => {
-    if (!columnMetrics) return [];
-
     return getViewportColumns(
-      columnMetrics.columns,
+      columns,
       colOverscanStartIdx,
       colOverscanEndIdx
     );
-  }, [colOverscanEndIdx, colOverscanStartIdx, columnMetrics]);
+  }, [colOverscanEndIdx, colOverscanStartIdx, columns]);
+
+  const totalHeaderHeight = headerRowHeight + (enableFilters ? headerFiltersHeight : 0);
+  const clientHeight = height
+    - 2 // border width
+    - totalHeaderHeight
+    - (summaryRows?.length ?? 0) * rowHeight
+    - (totalColumnWidth > viewportWidth ? getScrollbarSize() : 0);
 
+  const [rowOverscanStartIdx, rowOverscanEndIdx] = getVerticalRangeToRender(
+    clientHeight,
+    rowHeight,
+    scrollTop,
+    rows.length
+  );
+
+  /**
+   * effects
+   */
   useLayoutEffect(() => {
     // Do not calculate the width if width is provided
     if (width) return;
```

Comment:
I think we should use `typeof width === 'number'` instead as width can be 0 when using `Autosizer`

Reply:
Done